From 37a5e0ce7a69745d75029429059bfbdcdc3e2105 Mon Sep 17 00:00:00 2001
From: Taylor Neyland <tneyla@amazon.com>
Date: Wed, 19 Jul 2023 12:51:30 -0500
Subject: [PATCH 8/8] Networking improvements

- Disable UDP offload service for Redhat and Ubuntu
- Default Flatcar version to avoid pulling from internet on every make

Signed-off-by: Vignesh Goutham Ganesh <vgg@amazon.com>
---
 images/capi/Makefile                            |  2 +-
 .../system/disable-nic-offload-redhat.service   | 16 ++++++++++++++++
 .../system/disable-udp-offload-ubuntu.service   | 15 +++++++++++++++
 .../roles/providers/tasks/vmware-redhat.yml     | 17 +++++++++++++++++
 .../roles/providers/tasks/vmware-ubuntu.yml     | 17 +++++++++++++++++
 5 files changed, 66 insertions(+), 1 deletion(-)
 create mode 100644 images/capi/ansible/roles/providers/files/etc/systemd/system/disable-nic-offload-redhat.service
 create mode 100644 images/capi/ansible/roles/providers/files/etc/systemd/system/disable-udp-offload-ubuntu.service

diff --git a/images/capi/Makefile b/images/capi/Makefile
index 7174f0de3..e1f368fff 100644
--- a/images/capi/Makefile
+++ b/images/capi/Makefile
@@ -315,7 +315,7 @@ WINDOWS_VERSIONS		:=	windows-2019 windows-2019-efi windows-2022 windows-2022-efi
 
 # Set Flatcar Container Linux channel and version if not supplied
 FLATCAR_CHANNEL ?= stable
-FLATCAR_VERSION ?= current
+FLATCAR_VERSION ?= 3510.2.7
 ifeq ($(FLATCAR_VERSION),current)
 override FLATCAR_VERSION := $(shell hack/image-grok-latest-flatcar-version.sh $(FLATCAR_CHANNEL))
 endif
diff --git a/images/capi/ansible/roles/providers/files/etc/systemd/system/disable-nic-offload-redhat.service b/images/capi/ansible/roles/providers/files/etc/systemd/system/disable-nic-offload-redhat.service
new file mode 100644
index 000000000..3eb62c94b
--- /dev/null
+++ b/images/capi/ansible/roles/providers/files/etc/systemd/system/disable-nic-offload-redhat.service
@@ -0,0 +1,16 @@
+[Unit]
+Description=Disables UDP segmentation and TCP checksum offload
+After=NetworkManager-wait-online.service
+# Block manual interactions with this service
+RefuseManualStart=true
+RefuseManualStop=true
+
+[Service]
+Type=oneshot
+ExecStart=/usr/sbin/ethtool -K eth0 tx-udp_tnl-segmentation off
+ExecStart=/usr/sbin/ethtool -K eth0 tx-udp_tnl-csum-segmentation off
+ExecStart=/usr/sbin/ethtool -K eth0 rx off tx off
+RemainAfterExit=true
+
+[Install]
+WantedBy=multi-user.target
diff --git a/images/capi/ansible/roles/providers/files/etc/systemd/system/disable-udp-offload-ubuntu.service b/images/capi/ansible/roles/providers/files/etc/systemd/system/disable-udp-offload-ubuntu.service
new file mode 100644
index 000000000..7f5d50a8e
--- /dev/null
+++ b/images/capi/ansible/roles/providers/files/etc/systemd/system/disable-udp-offload-ubuntu.service
@@ -0,0 +1,15 @@
+[Unit]
+Description=Disables UDP offload for Ubuntu
+After=systemd-networkd-wait-online.service
+# Block manual interactions with this service
+RefuseManualStart=true
+RefuseManualStop=true
+
+[Service]
+Type=oneshot
+ExecStart=/usr/sbin/ethtool -K eth0 tx-udp_tnl-segmentation off
+ExecStart=/usr/sbin/ethtool -K eth0 tx-udp_tnl-csum-segmentation off
+RemainAfterExit=true
+
+[Install]
+WantedBy=network-online.target
\ No newline at end of file
diff --git a/images/capi/ansible/roles/providers/tasks/vmware-redhat.yml b/images/capi/ansible/roles/providers/tasks/vmware-redhat.yml
index bfb929a6e..2c881b2d0 100644
--- a/images/capi/ansible/roles/providers/tasks/vmware-redhat.yml
+++ b/images/capi/ansible/roles/providers/tasks/vmware-redhat.yml
@@ -68,3 +68,20 @@
       ansible.builtin.file:
         path: /tmp/cloud-init-vmware.sh
         state: absent
+
+- name: Create service disable udp segmentation offload and tcp checksum offload
+  copy:
+    src: files/etc/systemd/system/disable-nic-offload-redhat.service
+    dest: /etc/systemd/system/disable-nic-offload-redhat.service
+    owner: root
+    group: root
+    mode: 0644
+  when: ansible_os_family != "Flatcar"
+
+- name: Enable disable-nic-offload-redhat.service
+  systemd:
+    name: disable-nic-offload-redhat.service
+    daemon_reload: yes
+    enabled: True
+    state: stopped
+  when: ansible_os_family != "Flatcar"
diff --git a/images/capi/ansible/roles/providers/tasks/vmware-ubuntu.yml b/images/capi/ansible/roles/providers/tasks/vmware-ubuntu.yml
index 46e524ee0..4e3f800b1 100644
--- a/images/capi/ansible/roles/providers/tasks/vmware-ubuntu.yml
+++ b/images/capi/ansible/roles/providers/tasks/vmware-ubuntu.yml
@@ -62,3 +62,20 @@
     content: |
       datasource: VMware
   when: ansible_distribution_version is version('22.04', '>=')
+
+- name: Create service disable udp offload
+  copy:
+    src: files/etc/systemd/system/disable-udp-offload-ubuntu.service
+    dest: /etc/systemd/system/disable-udp-offload-ubuntu.service
+    owner: root
+    group: root
+    mode: 0644
+  when: ansible_distribution_version is version('22.04', '>=')
+
+- name: Enable disable-udp-offload-ubuntu.service
+  systemd:
+    name: disable-udp-offload-ubuntu.service
+    daemon_reload: yes
+    enabled: True
+    state: stopped
+  when: ansible_distribution_version is version('22.04', '>=')
-- 
2.45.2

